File: turboci/id/wrap.go
// Copyright 2025 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package id

import (
	"fmt"

	idspb "go.chromium.org/turboci/proto/go/graph/ids/v1"
)

// StageIdentifier is the union of all valid Identifier messages which are
// rooted via a Stage in the graph.
type StageIdentifier interface {
	*idspb.Identifier |
		*idspb.Stage |
		*idspb.StageAttempt |
		*idspb.StageEdit |
		*idspb.StageEditReason
}

// CheckIdentifier is the union of all valid Identifier messages which are
// rooted via a Check in the graph.
type CheckIdentifier interface {
	*idspb.Identifier |
		*idspb.Check |
		*idspb.CheckOption |
		*idspb.CheckResult |
		*idspb.CheckResultDatum |
		*idspb.CheckEdit |
		*idspb.CheckEditReason |
		*idspb.CheckEditOption
}

// Identifier is the union of all valid Identifier messages.
type Identifier interface {
	*idspb.Identifier |
		*idspb.WorkPlan |
		StageIdentifier |
		CheckIdentifier
}

// Wrap takes any Identifier sub-type and wraps it into a generic
// idspb.Identifier.
func Wrap[Id Identifier](id Id) *idspb.Identifier {
	return wrap(id)
}

// wrap takes any Identifier type and returns it as an idspb.Identifier.
func wrap(id any) *idspb.Identifier {
	switch x := id.(type) {
	case nil:
		return nil
	case *idspb.Identifier:
		return x
	case *idspb.WorkPlan:
		return idspb.Identifier_builder{WorkPlan: x}.Build()
	case *idspb.Check:
		return idspb.Identifier_builder{Check: x}.Build()
	case *idspb.CheckOption:
		return idspb.Identifier_builder{CheckOption: x}.Build()
	case *idspb.CheckResult:
		return idspb.Identifier_builder{CheckResult: x}.Build()
	case *idspb.CheckResultDatum:
		return idspb.Identifier_builder{CheckResultDatum: x}.Build()
	case *idspb.CheckEdit:
		return idspb.Identifier_builder{CheckEdit: x}.Build()
	case *idspb.CheckEditReason:
		return idspb.Identifier_builder{CheckEditReason: x}.Build()
	case *idspb.CheckEditOption:
		return idspb.Identifier_builder{CheckEditOption: x}.Build()
	case *idspb.Stage:
		return idspb.Identifier_builder{Stage: x}.Build()
	case *idspb.StageAttempt:
		return idspb.Identifier_builder{StageAttempt: x}.Build()
	case *idspb.StageEdit:
		return idspb.Identifier_builder{StageEdit: x}.Build()
	case *idspb.StageEditReason:
		return idspb.Identifier_builder{StageEditReason: x}.Build()
	}

	panic(fmt.Sprintf("bug: id.wrap got unexpected type %T", id))
}

// unwraps any Identifier into it's *specific* identifier type.
//
// Allows generic functions to be written with a simpler type switch.
func unwrap(id any) any {
	if id == nil {
		return nil
	}

	switch x := id.(type) {
	case *idspb.Identifier:
		switch x.WhichType() {
		case idspb.Identifier_WorkPlan_case:
			return x.GetWorkPlan()
		case idspb.Identifier_Check_case:
			return x.GetCheck()
		case idspb.Identifier_CheckOption_case:
			return x.GetCheckOption()
		case idspb.Identifier_CheckResult_case:
			return x.GetCheckResult()
		case idspb.Identifier_CheckResultDatum_case:
			return x.GetCheckResultDatum()
		case idspb.Identifier_CheckEdit_case:
			return x.GetCheckEdit()
		case idspb.Identifier_CheckEditOption_case:
			return x.GetCheckEditOption()
		case idspb.Identifier_CheckEditReason_case:
			return x.GetCheckEditReason()
		case idspb.Identifier_Stage_case:
			return x.GetStage()
		case idspb.Identifier_StageAttempt_case:
			return x.GetStageAttempt()
		case idspb.Identifier_StageEdit_case:
			return x.GetStageEdit()
		case idspb.Identifier_StageEditReason_case:
			return x.GetStageEditReason()
		case idspb.Identifier_Type_not_set_case:
			return nil
		}

	case *idspb.WorkPlan,
		*idspb.Stage,
		*idspb.StageAttempt,
		*idspb.StageEdit,
		*idspb.StageEditReason,
		*idspb.Check,
		*idspb.CheckOption,
		*idspb.CheckResult,
		*idspb.CheckResultDatum,
		*idspb.CheckEdit,
		*idspb.CheckEditReason,
		*idspb.CheckEditOption:

		return x
	}

	panic(fmt.Sprintf("bug: id.unwrap got unexpected type %T", id))
}

// StageRoot returns the *Stage for any StageIdentifier.
//
// If the Stage portion of the identifier is `nil`, this returns `nil`.
func StageRoot[Id StageIdentifier](id Id) *idspb.Stage {
	_, _, stage := Root(id)
	return stage
}

// CheckRoot returns the *Check for any CheckIdentifier.
//
// If the Stage portion of the identifier is `nil`, this returns `nil`.
func CheckRoot[Id CheckIdentifier](id Id) *idspb.Check {
	_, check, _ := Root(id)
	return check
}

// Root returns either the Check or Stage which serves as the root for the
// specific id, plus the WorkPlan.
//
// For *idspb.WorkPlan (or a wrapped WorkPlan), this just returns `wp`.
//
// If the check/stage portion of `id` is nil, this returns nil.
// If the WorkPlan portion of `id` is nil, this returns nil.
func Root[Id Identifier](id Id) (wp *idspb.WorkPlan, check *idspb.Check, stage *idspb.Stage) {
	switch x := unwrap(id).(type) {
	case *idspb.WorkPlan:
		wp = x

	// Stage root
	case *idspb.Stage:
		stage = x
	case *idspb.StageAttempt:
		stage = x.GetStage()
	case *idspb.StageEdit:
		stage = x.GetStage()
	case *idspb.StageEditReason:
		stage = x.GetStageEdit().GetStage()

	// Check root
	case *idspb.Check:
		check = x
	case *idspb.CheckOption:
		check = x.GetCheck()
	case *idspb.CheckResult:
		check = x.GetCheck()
	case *idspb.CheckResultDatum:
		check = x.GetResult().GetCheck()
	case *idspb.CheckEdit:
		check = x.GetCheck()
	case *idspb.CheckEditReason:
		check = x.GetCheckEdit().GetCheck()
	case *idspb.CheckEditOption:
		check = x.GetCheckEdit().GetCheck()
	case nil:
		// nil identifiers have check, stage and workplan all-nil.

	default:
		panic(fmt.Sprintf("bug: id.Root got unexpected type %T", id))
	}

	// Fill in WorkPlan from the stage/check.
	if stage != nil {
		wp = stage.GetWorkPlan()
	} else if check != nil {
		wp = check.GetWorkPlan()
	}

	return
}

// KindOf describes any type of Identifier as an IdentifierKind enum.
func KindOf[Id Identifier](id Id) idspb.IdentifierKind {
	switch typ := unwrap(id); typ {
	case idspb.Identifier_WorkPlan_case:
		return idspb.IdentifierKind_IDENTIFIER_KIND_WORK_PLAN
	case idspb.Identifier_Check_case:
		return idspb.IdentifierKind_IDENTIFIER_KIND_CHECK
	case idspb.Identifier_CheckOption_case:
		return idspb.IdentifierKind_IDENTIFIER_KIND_CHECK_OPTION
	case idspb.Identifier_CheckResult_case:
		return idspb.IdentifierKind_IDENTIFIER_KIND_CHECK_RESULT
	case idspb.Identifier_CheckResultDatum_case:
		return idspb.IdentifierKind_IDENTIFIER_KIND_CHECK_RESULT_DATUM
	case idspb.Identifier_CheckEdit_case:
		return idspb.IdentifierKind_IDENTIFIER_KIND_CHECK_EDIT
	case idspb.Identifier_CheckEditReason_case:
		return idspb.IdentifierKind_IDENTIFIER_KIND_CHECK_EDIT_REASON
	case idspb.Identifier_CheckEditOption_case:
		return idspb.IdentifierKind_IDENTIFIER_KIND_CHECK_EDIT_OPTION
	case idspb.Identifier_Stage_case:
		return idspb.IdentifierKind_IDENTIFIER_KIND_STAGE
	case idspb.Identifier_StageAttempt_case:
		return idspb.IdentifierKind_IDENTIFIER_KIND_STAGE_ATTEMPT
	case idspb.Identifier_StageEdit_case:
		return idspb.IdentifierKind_IDENTIFIER_KIND_STAGE_EDIT
	case idspb.Identifier_StageEditReason_case:
		return idspb.IdentifierKind_IDENTIFIER_KIND_CHECK_EDIT_REASON

	case idspb.Identifier_Type_not_set_case:
		return idspb.IdentifierKind_IDENTIFIER_KIND_UNKNOWN

	default:
		panic(fmt.Sprintf("impossible type: %s", typ))
	}
}

File: milo/ui/src/chronicle/utils/id/wrap.ts
// Copyright 2026 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {
  Check,
  CheckEdit,
  CheckResult,
  Identifier,
  Stage,
  StageAttempt,
  StageEdit,
  WorkPlan,
} from '@/proto/turboci/graph/ids/v1/identifier.pb';
import { IdentifierKind } from '@/proto/turboci/graph/ids/v1/identifier_kind.pb';

export type ConcreteIdentifier =
  | WorkPlan
  | Check
  | CheckResult
  | CheckEdit
  | Stage
  | StageAttempt
  | StageEdit;

/**
 * Type guard to identify if an object is an Identifier container.
 */
export function isIdentifier(obj: unknown): obj is Identifier {
  if (!obj || typeof obj !== 'object') return false;
  const record = obj as Record<string, unknown>;
  if ('id' in record || 'idx' in record || 'version' in record) return false;
  return (
    'workPlan' in record ||
    'check' in record ||
    'checkResult' in record ||
    'checkEdit' in record ||
    'stage' in record ||
    'stageAttempt' in record ||
    'stageEdit' in record
  );
}

export function isWorkPlan(obj: unknown): obj is WorkPlan {
  if (!obj || typeof obj !== 'object') return false;
  const record = obj as Record<string, unknown>;
  return (
    'id' in record &&
    !(
      'workPlan' in record ||
      'idx' in record ||
      'version' in record ||
      'isWorknode' in record
    )
  );
}

export function isCheck(obj: unknown): obj is Check {
  if (!obj || typeof obj !== 'object') return false;
  const record = obj as Record<string, unknown>;
  return (
    'id' in record &&
    'workPlan' in record &&
    !('idx' in record || 'version' in record || 'isWorknode' in record)
  );
}

export function isStage(obj: unknown): obj is Stage {
  if (!obj || typeof obj !== 'object') return false;
  const record = obj as Record<string, unknown>;
  return (
    'id' in record &&
    'workPlan' in record &&
    'isWorknode' in record &&
    !('idx' in record || 'version' in record)
  );
}

export function isCheckResult(obj: unknown): obj is CheckResult {
  if (!obj || typeof obj !== 'object') return false;
  const record = obj as Record<string, unknown>;
  return 'idx' in record && 'check' in record;
}

export function isStageAttempt(obj: unknown): obj is StageAttempt {
  if (!obj || typeof obj !== 'object') return false;
  const record = obj as Record<string, unknown>;
  return 'idx' in record && 'stage' in record;
}

export function isCheckEdit(obj: unknown): obj is CheckEdit {
  if (!obj || typeof obj !== 'object') return false;
  const record = obj as Record<string, unknown>;
  return 'version' in record && 'check' in record;
}

export function isStageEdit(obj: unknown): obj is StageEdit {
  if (!obj || typeof obj !== 'object') return false;
  const record = obj as Record<string, unknown>;
  return 'version' in record && 'stage' in record;
}

export function isConcreteIdentifier(obj: unknown): obj is ConcreteIdentifier {
  return (
    isWorkPlan(obj) ||
    isCheck(obj) ||
    isCheckResult(obj) ||
    isCheckEdit(obj) ||
    isStage(obj) ||
    isStageAttempt(obj) ||
    isStageEdit(obj)
  );
}

/**
 * Wraps any specific Identifier sub-type into a generic Identifier container.
 */
export function wrap(id: unknown): Identifier | null {
  if (!id) return null;
  if (isIdentifier(id)) return id;

  if (isWorkPlan(id)) return { workPlan: id };
  if (isCheck(id)) return { check: id };
  if (isCheckResult(id)) return { checkResult: id };
  if (isCheckEdit(id)) return { checkEdit: id };
  if (isStage(id)) return { stage: id };
  if (isStageAttempt(id)) return { stageAttempt: id };
  if (isStageEdit(id)) return { stageEdit: id };

  return null;
}

/**
 * Unwraps a generic Identifier container into its concrete underlying type.
 */
export function unwrap(id: unknown): ConcreteIdentifier | null {
  if (!id) return null;
  if (isIdentifier(id)) {
    if (id.workPlan !== undefined) return id.workPlan;
    if (id.check !== undefined) return id.check;
    if (id.checkResult !== undefined) return id.checkResult;
    if (id.checkEdit !== undefined) return id.checkEdit;
    if (id.stage !== undefined) return id.stage;
    if (id.stageAttempt !== undefined) return id.stageAttempt;
    if (id.stageEdit !== undefined) return id.stageEdit;
    return null;
  }
  if (isConcreteIdentifier(id)) {
    return id;
  }
  return null;
}

/**
 * Resolves the stage root of a StageIdentifier.
 */
export function stageRoot(id: unknown): Stage | null {
  const r = root(id);
  return r.stage;
}

/**
 * Resolves the check root of a CheckIdentifier.
 */
export function checkRoot(id: unknown): Check | null {
  const r = root(id);
  return r.check;
}

/**
 * Resolves the base check/stage and workplan associated with any ID container or concrete ID object.
 */
export function root(id: unknown): {
  wp: WorkPlan | null;
  check: Check | null;
  stage: Stage | null;
} {
  let wp: WorkPlan | null = null;
  let check: Check | null = null;
  let stage: Stage | null = null;

  if (isIdentifier(id)) {
    if (id.workPlan !== undefined) {
      wp = id.workPlan;
    } else if (id.check !== undefined) {
      check = id.check;
    } else if (id.checkResult !== undefined) {
      check = id.checkResult.check || null;
    } else if (id.checkEdit !== undefined) {
      check = id.checkEdit.check || null;
    } else if (id.stage !== undefined) {
      stage = id.stage;
    } else if (id.stageAttempt !== undefined) {
      stage = id.stageAttempt.stage || null;
    } else if (id.stageEdit !== undefined) {
      stage = id.stageEdit.stage || null;
    }
  } else {
    const unwrapped = unwrap(id);
    if (!unwrapped) {
      return { wp, check, stage };
    }

    if (isWorkPlan(unwrapped)) {
      wp = unwrapped;
    } else if (isCheck(unwrapped)) {
      check = unwrapped;
    } else if (isStage(unwrapped)) {
      stage = unwrapped;
    } else if (isCheckResult(unwrapped)) {
      check = unwrapped.check || null;
    } else if (isStageAttempt(unwrapped)) {
      stage = unwrapped.stage || null;
    } else if (isCheckEdit(unwrapped)) {
      check = unwrapped.check || null;
    } else if (isStageEdit(unwrapped)) {
      stage = unwrapped.stage || null;
    }
  }

  // Fill in WorkPlan from check or stage if possible
  if (stage) {
    wp = stage.workPlan || null;
  } else if (check) {
    wp = check.workPlan || null;
  }

  return { wp, check, stage };
}

/**
 * Describes any type of Identifier as an IdentifierKind enum.
 */
export function kindOf(id: unknown): IdentifierKind {
  if (isIdentifier(id)) {
    if (id.workPlan !== undefined)
      return IdentifierKind.IDENTIFIER_KIND_WORK_PLAN;
    if (id.check !== undefined) return IdentifierKind.IDENTIFIER_KIND_CHECK;
    if (id.checkResult !== undefined)
      return IdentifierKind.IDENTIFIER_KIND_CHECK_RESULT;
    if (id.checkEdit !== undefined)
      return IdentifierKind.IDENTIFIER_KIND_CHECK_EDIT;
    if (id.stage !== undefined) return IdentifierKind.IDENTIFIER_KIND_STAGE;
    if (id.stageAttempt !== undefined)
      return IdentifierKind.IDENTIFIER_KIND_STAGE_ATTEMPT;
    if (id.stageEdit !== undefined)
      return IdentifierKind.IDENTIFIER_KIND_STAGE_EDIT;
  }

  const unwrapped = unwrap(id);
  if (!unwrapped) {
    throw new Error(`unknown or invalid identifier: ${id}`);
  }

  if (isWorkPlan(unwrapped)) return IdentifierKind.IDENTIFIER_KIND_WORK_PLAN;
  if (isCheck(unwrapped)) return IdentifierKind.IDENTIFIER_KIND_CHECK;
  if (isCheckResult(unwrapped))
    return IdentifierKind.IDENTIFIER_KIND_CHECK_RESULT;
  if (isCheckEdit(unwrapped)) return IdentifierKind.IDENTIFIER_KIND_CHECK_EDIT;
  if (isStage(unwrapped)) return IdentifierKind.IDENTIFIER_KIND_STAGE;
  if (isStageAttempt(unwrapped))
    return IdentifierKind.IDENTIFIER_KIND_STAGE_ATTEMPT;
  if (isStageEdit(unwrapped)) return IdentifierKind.IDENTIFIER_KIND_STAGE_EDIT;

  throw new Error(`impossible type for kindOf: ${JSON.stringify(unwrapped)}`);
}

/**
 * Returns true if the two identifiers have the same root.
 */
export function sameRoot(a: unknown, b: unknown): boolean {
  if (!a || !b) {
    return false;
  }
  const aRoot = root(a);
  const bRoot = root(b);

  if (aRoot.stage && bRoot.stage) {
    return (
      aRoot.stage.id === bRoot.stage.id &&
      aRoot.stage.isWorknode === bRoot.stage.isWorknode &&
      aRoot.stage.workPlan?.id === bRoot.stage.workPlan?.id
    );
  }

  if (!aRoot.stage && !bRoot.stage) {
    if (!aRoot.check && !bRoot.check) {
      return true;
    }
    if (aRoot.check && bRoot.check) {
      return (
        aRoot.check.id === bRoot.check.id &&
        aRoot.check.workPlan?.id === bRoot.check.workPlan?.id
      );
    }
  }

  return false;
}

/**
 * Returns true if the two identifiers have the same workplan.
 */
export function sameWorkPlan(a: unknown, b: unknown): boolean {
  if (!a || !b) {
    return false;
  }
  const aRoot = root(a);
  const bRoot = root(b);

  if (!aRoot.wp && !bRoot.wp) {
    return true;
  }
  if (aRoot.wp && bRoot.wp) {
    return aRoot.wp.id === bRoot.wp.id;
  }
  return false;
}
